### ISSUE TITLE
@QueryInit not working with deep nested paths

### ISSUE THREAD

**Issue by manish-in-java** [querydsl/querydsl#2129]
I am using QueryDSL 4.1.4 with Spring Data in a project and am having trouble using it with nested paths. Consider the following JPA classes:

    @Entity @QueryEntity class Company {}

    @Entity
    @QueryEntity
    class Contract {
      @ManyToOne
      @QueryInit("*.*")
      Company buyer;

      @ManyToOne
      @QueryInit("*.*")
      Company seller;
    }

    @Entity @QueryEntity class Bill { @ManyToOne @QueryInit("*.*") Contract contract; }

    @Entity @QueryEntity class Payment { @ManyToOne @QueryInit("*.*") Bill bill; }

    @Entity @QueryEntity class Posting { @ManyToOne @QueryInit("*.*") Payment payment; }

With this set up, the following throws `NullPointerException`:

    QPosting.posting.payment.bill.contract.buyer.name.equalsIgnoreCase(...)

Debugging shows that `QPosting.posting.payment.bill.contract.buyer` (and `QPosting.posting.payment.bill.contract.seller`) is `null`.

The following however work fine:

1. `QPayment.payment.bill.contract.buyer.name.equalsIgnoreCase(...)`
1. `QBill.bill.contract.buyer.name.equalsIgnoreCase(...)`
1. `QContract.contract.buyer.name.equalsIgnoreCase(...)`

I am using `com.mysema.maven:apt-maven-plugin:1.1.1` and QueryDSL version 4.1.4.

A full sample project is available [here](https://github.com/manish-in-java/querydsl-nested-path). Run Maven tests as `mvn clean test` to see the error.

**Comment by marceloverdijk** [comment]
I'm experiencing the same issue.

I cannot actually point a finger to why some nested paths are `null` but the strange thing is that it is not the case for all entities.

E.g I have paths like:

```
        Object o = QDriver.driver.nationality.continent;
        Object p = QDriver.driver.nationality.continent.code;
        Object q = QDriver.driver.nationality.continent.name;
        Object r = QDriver.driver.secondNationality.continent;
        Object s = QDriver.driver.secondNationality.continent.code;
        Object t = QDriver.driver.secondNationality.continent.name;
```

which works perfectly, but I also have:

```
        Object o = QRace.race.circuit.country;
        Object p = QRace.race.circuit.country.continent;
        Object q = QRace.race.circuit.country.continent.code;
        Object r = QRace.race.circuit.country.continent.name;
```

where the `QRace.race.circuit.country.continent` is `null` and causing a `NPE` accessing the `code` field.

The more interesting is that 

```
        Object o = QCircuit.circuit.country.name;
        Object p = QCircuit.circuit.country.continent.code;
        Object q = QCircuit.circuit.country.continent.name;
``

works without problems which are under the covers the same entity classes.

Is there any explanation why this is failing or a possible workaround?

**Comment by FranPregernik** [comment]
I just encountered the same issue in my model.

I get a `NullPointerException` when calling `upisaniPredmet.upisaniElement.upis.student`.

`upisaniPredmet.upisaniElement` is not null.
`upisaniPredmet.upisaniElement.upis` is not null.
`upisaniPredmet.upisaniElement.upis.student` **is** null.

**Comment by mosekk** [comment]
I have exactly the same problem. Must be the bug

**Comment by felipexw** [comment]
+1

**Comment by POJOa** [comment]
any solutions? facing it too.

**Comment by spyro2000** [comment]
Same here, expected this just to work out of the box :(

**Comment by JoeToffler** [comment]
Please give us the answer,thank you verymuch.

**Comment by seongjin605** [comment]
@POJOa Thank you very much :)

**Comment by lucasfrederico** [comment]
**Someone managed to resolve the error? I have the same problem.**

**Comment by seongjin605** [comment]
@lucasfrederico 
By default Querydsl initializes only reference properties of the first two levels. In cases where longer initialization paths are required, these have to be annotated in the domain types via com.mysema.query.annotations.QueryInit annotations. QueryInit is used on properties where deep initializations are needed.

http://www.querydsl.com/static/querydsl/3.5.0/reference/html/ch03s03.html#d0e2181

**Comment by rnferreira** [comment]
Hey, sorry to bump this old thread but I'm facing the same problem. Is there a fix for this?

**Comment by jwgmeligmeyling** [comment]
There is no fix for this, and it will likely not get fixed because initializing paths so deep is incredibly computationally expensive. As a result, QueryInits from referenced classes are not used transitively, so you have to specify the QueryInits where you want them.

If you want to initialise the path `QPosting.posting.payment.bill.contract.buyer` you will explicitly have to initialise it, as such:

```java
@Entity @QueryEntity class Posting {
    @ManyToOne @QueryInit("*.*", "bill.contract.buyer") Payment payment;
}
```

That said, there is one known issue regarding path initialisation, and that one is related to entities that implement `Comparable`. That issue is described in #2572 and I have suggested a fix in  #2573.
